File: WolfGoatCabbage.js
function isValidState(state) {
    const [farmer, wolf, goat, cabbage] = state;
    if (wolf === goat && farmer !== wolf) return false;
    if (goat === cabbage && farmer !== goat) return false;
    return true;
}

function getNextStates(currentState) {
    const nextStates = [];
    const [farmer, wolf, goat, cabbage] = currentState;

    for (let i = 0; i < 4; i++) {
        if (i === 0 || currentState[i] === farmer) {
            const newState = [...currentState];
            newState[0] = 1 - farmer; // Перевозимо фермера
            if (i !== 0) {
                newState[i] = 1 - currentState[i]; // Перевозимо об'єкт
            }
            if (isValidState(newState)) {
                nextStates.push(newState);
            }
        }
    }

    return nextStates;
}

function findSolution() {
    const startState = [0, 0, 0, 0]; // Фермер, вовк, коза, капуста
    const goalState = [1, 1, 1, 1]; // Усі на іншому березі
    const visited = new Set();
    const queue = [[startState, []]]; // Стан та шлях до нього

    while (queue.length > 0) {
        const [currentState, path] = queue.shift();
        if (currentState.toString() === goalState.toString()) {
            return [...path, currentState];
        }

        getNextStates(currentState).forEach(nextState => {
            const stateString = nextState.toString();
            if (!visited.has(stateString)) {
                visited.add(stateString);
                queue.push([nextState, [...path, currentState]]);
            }
        });
    }

    return null;
}

// Знаходимо рішення
const solution = findSolution();
console.log(solution);

File: WolfGoatCabbage.pl
% State is described as state(Farmer, Wolf, Goat, Cabbage),
% where each variable can be either 'left' or 'right'.

% Checking if a state is safe.
safe_state(state(F, W, G, C)) :-
    \+ unsafe(F, W, G, C).

unsafe(F, W, G, C) :-
    (F \= W, W = G);  % Wolf and Goat are on the same side, Farmer is not
    (F \= G, G = C).  % Goat and Cabbage are on the same side, Farmer is not

% Moving the farmer alone or with an animal.
move(state(X, W, G, C), state(Y, W, G, C)) :- opposite(X, Y).
move(state(X, X, G, C), state(Y, Y, G, C)) :- opposite(X, Y).
move(state(X, W, X, C), state(Y, W, Y, C)) :- opposite(X, Y).
move(state(X, W, G, X), state(Y, W, G, Y)) :- opposite(X, Y).

% Defining the opposite side.
opposite(left, right).
opposite(right, left).

% Recursive solution search.
solution(State, Visited, Path) :-
    target_state(State),
    reverse(Visited, Path).

solution(State, Visited, Path) :-
    move(State, NextState),
    safe_state(NextState),
    \+ member(NextState, Visited),  % Ensure the state has not been visited before
    solution(NextState, [NextState | Visited], Path).

% Defining the target state.
target_state(state(right, right, right, right)).

% Initiating the solution search.
find_solution(Solution) :-
    solution(state(left, left, left, left), [state(left, left, left, left)], Solution).
